[3.x] Backport Add support for SDL3 joystick input driver

kilian-diener opened the pull request:
Based on #106218 and #108373

Backports the SDL input driver to 3.x. It follows closely the current implementation on Master.
I built and verified the changes on Linux, Windows and Mac with the Joypad Demo project and the following controllers: PS4, PS5, Switch Pro, Xbox Series X (1914), Xbox One S (1708) and Nacon off brand Xbox 360.

- [X] : How to correctly handle udev and dbus on Linux? SOWARP_ENABLED is not present on 3.x but there is a libudev-so_wrap. Is a libdbus-so_wrap required to make it work if dbus can't be found on the system? Do we need a thirdparty/linuxbsd_headers as well? I'm a bit out of my depth with understanding how and when which libraries is used from where.

kilian-diener committed:
Backport Add support for SDL3 joystick input driver

Made possible by EIREXE, xsellier and the SDL team.

This commit includes statically linked SDL3 for Windows, Linux and macOS.
The vendored copy of SDL3 was setup to only build the required subsystems
for gamepad/joystick support, with some patches to be able to make it as
minimal as possible and reduce the impact on binary size and code size.

Co-authored-by: Nintorch <92302738+Nintorch@users.noreply.github.com>
Co-authored-by: Álex Román Núñez <eirexe123@gmail.com>
Co-authored-by: Xavier Sellier <xsellier@gmail.com>
Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

kilian-diener commented:
Updated with the fixes from #109750 #110198 #109819 #109283

kilian-diener committed:
Backport #111192 Fix invalid reported joypad presses

Co-authored-by: Nintorch <92302738+Nintorch@users.noreply.github.com>

kilian-diener committed:
Backport #111191 Fix weak and strong joypad vibration being swapped

Co-authored-by: Nintorch <92302738+Nintorch@users.noreply.github.com>

kilian-diener committed:
Backport #111146 Fix compiling SDL without DBus under Linux

Co-authored-by: Kaleb Reid <78945904+Kaleb-Reid@users.noreply.github.com>

AThousandShips commented on drivers/sdl/joypad_sdl.cpp:
```diff
@@ -0,0 +1,367 @@
+// Required for DInput joypads to work
+// TODO: remove this workaround when we update to newer version of SDL
+JoypadSDL::JoypadSDL(InputDefault *in, HWND p_helper_window) :
+		JoypadSDL(in) {
```
```suggestion
JoypadSDL::JoypadSDL(InputDefault *p_in, HWND p_helper_window) :
		JoypadSDL(p_in) {
```

kilian-diener commented:
Did some measurements on my laptop: Intel Core Ultra 9 185H (22 threads) - 30.73 GiB memory - Nvidia 4070 Laptop - Arch 6.17.2 with GCC 15.2.1
Editor command: scons platform=x11/windows target=release_debug
Template command: scons platform=x11/windows tools=no target=release bits=64
Inbetween each run did the same command with -c to clean up.

Linux:
|         | Editor Build Time | Editor Size | Editor Size Stripped | Template Build Time | Template Size | Template Size Stripped |
|---------|-------------|-------------|----------------------|---------------|---------------|------------------------|
| 3.x     |        5:18 |       889.1MB |                 71.5MB |          3:23 |         514.3MB |                   39.7MB |
| 3.x SDL |        5:25 |       891.2MB |                   72MB |          3:27 |         516.7MB |                   40.2MB |
| Difference |        0:07 |       2.1MB |                0.5MB |          0:04 |         2.4MB |                  0.5MB |

Windows:
|         | Editor Build Time | Editor Size | Editor Size Stripped | Template Build Time | Template Size | Template Size Stripped |
|---------|-------------|-------------|----------------------|---------------|---------------|------------------------|
| 3.x     |        5:19 |       939.5MB |                 71.7MB |          3:22 |           530MB |                   38.8MB |
| 3.x SDL |        5:25 |       942.9MB |                 72.2MB |          3:27 |         534.5MB |                   39.4MB |
| Difference |        0:06 |       3.4MB |                0.5MB |          0:05 |         4.5MB |                  0.6MB |

Let me know if you need more information.

kilian-diener committed:
Fix print_warning not existing on 3.x. Use print instead.

kilian-diener committed:
Backport #113968 sdl: Update to 3.2.28

AThousandShips commented on drivers/sdl/joypad_sdl.cpp:
```diff
@@ -0,0 +1,354 @@
+
+#ifdef SDL_ENABLED
+
+#include "core/dictionary.h"
```
```suggestion
```

Nintorch commented:
This PR should likely fix these issues for 3.x since they've been reported as fixed for 4.5.1:
https://github.com/godotengine/godot/issues/66878
https://github.com/godotengine/godot/issues/45443

kilian-diener commented:
I'll backport HIDAPI and check out other changes to master to see if something makes sense to be backported as well

kilian-diener committed:
Backport HIDAPI support from #111707

Co-authored-by: Nintorch <92302738+Nintorch@users.noreply.github.com>

kilian-diener committed:
Backport #116544: [CODEOWNERS] Add codeowners for drivers/sdl/

This PR makes the Input team a codeowner for `drivers/sdl/` folder, since at the moment SDL is only used for joypad input.
If in the future we add more subsystems from SDL, we can probably change the codeowners too.
Co-authored-by: Nintorch <92302738+Nintorch@users.noreply.github.com>

kilian-diener committed:
Backport #115952: Remove unused JoypadSDL::singleton

Co-authored-by: Nintorch <92302738+Nintorch@users.noreply.github.com>

kilian-diener committed:
Backport #115510: Fix Nvidia Shield controller showing up twice on macOS

Co-authored-by: Nintorch <92302738+Nintorch@users.noreply.github.com>

kilian-diener committed:
Take over minor changes to joypad_sdl.cpp to better reflect master
